src/iov.c

```diff
@@ -101,3 +101,41 @@ int ofi_truncate_iov(struct iovec *iov, size_t *iov_count, size_t trim_size)
 	}
 	return -FI_ETRUNC;
 }
+
+/* Move 'len' bytes worth of scatter / gather entries and desc from src to dst */
+int ofi_move_iov_desc(struct ofi_iov_desc *dst, struct ofi_iov_desc *src, size_t len)
+{
+	size_t i;
+
+	for (i = 0; (i < src->count) && len; i++) {
+		dst->iov[i].iov_base = src->iov[i].iov_base;
+		if (src->desc)
+			dst->desc[i] = src->desc[i];
+
+		if (len < src->iov[i].iov_len) {
+			dst->iov[i].iov_len = len;
+			break;
+		} else {
+			dst->iov[i].iov_len = src->iov[i].iov_len;
+		}
+
+		len -= dst->iov[i].iov_len;
+	}
+
+	if (OFI_UNLIKELY((i == src->count) && len))
+		return -FI_ETOOSMALL;
+
+	dst->count = len ? i + 1 : i;
+	src->count -= i;
+
+	if (src->count) {
+		memmove(src->iov, &src->iov[i], src->count * sizeof(*src->iov));
+		if (src->desc)
+			memmove(src->desc, &src->desc[i],
+				src->count * sizeof(*src->desc));
+
+		src->iov[0].iov_base += len;
+		src->iov[0].iov_len -= len;
+	}
+	return 0;
+}
```

Comment:
This looks like a merger of memcpy iovecs, ofi_consume_iov(), followed by ofi_truncate_iov().

Reply:
Yes. In addition, this avoids looping through src->iov again (since consume and truncate are merged into one) and also handles desc.